### PR TITLE
first pass at dockerized testing

### PR DESCRIPTION
**Pull request by lrvick**
Do not merge.

3 tests still not passing, but comments welcome

Rationale here is tests kept failing due to custom gpg/ssh/env settings etc and then I realized testing on the host system is a terrible idea so refactored things to test in a container.

Wanting to use this as a base to write out some tests around a fix for our 'login' issues.

In general trying to debug bats is a real pita so I added 'make develop' which drops you into the docker-composed testing env with test_helpers sourced so you can step through failing tests REPL style.

### CODE REVIEW

**Comment by lrvick** [comment]
Any ideas on making this play nice with Travis? 

Suppose I can make an alternate test-native if needed.

**Comment by DeviaVir** [comment]
Don't want to steal @thepwagner's spotlight here, but why not switch to jenks (see: https://github.com/pebble/spacel-agent/blob/master/Jenkinsfile)

**Comment by lrvick** [comment]
Okay, not a clue why 7 tests fail in travis, but not outside of it. Considering this is a public project travis seems ideal, but only if it works...

If someone else could manually run tests and confirm this works as expected on OSX that would be cool.

**Comment by DeviaVir** [comment]
```
...snip... (all same errors)
 ✗ Run script from hook dir - not found
   (from function `destroy_backups' in file test/test_helper.bash, line 2,
    from function `setup' in test file test/test.bats, line 6)
     `destroy_backups' failed
 ✗ Run script from hook dir - no pull
   (from function `destroy_backups' in file test/test_helper.bash, line 2,
    from function `setup' in test file test/test.bats, line 6)
     `destroy_backups' failed

34 tests, 34 failures
```
On OSX

**Comment by lrvick** [comment]
@DeviaVir what version of docker? docker-compose? Does "make develop" work?

This setup is little more than a docker-compose file and containers as far as your OS -should- be concerned.

**Comment by DeviaVir** [comment]
```
Chase ~ git-deploy $ docker version
Client:
 Version:      1.12.0-rc3
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   91e29e8
 Built:        Sat Jul  2 00:09:24 2016
 OS/Arch:      darwin/amd64
 Experimental: true

Server:
 Version:      1.12.0-rc3
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   876f3a7
 Built:        Tue Jul  5 02:20:13 2016
 OS/Arch:      linux/amd64
 Experimental: true


Chase ~ git-deploy $ docker-compose --version
docker-compose version 1.8.0-rc1, build 9bf6bc6


Chase ~ git-deploy $ git-deploy  make develop
docker rm -f gitdeploy_git-deploy_1 || return 0
Error response from daemon: No such container: gitdeploy_git-deploy_1
/bin/sh: line 0: return: can only `return' from a function or sourced script
make: *** [clean] Error 1
```

**Comment by lrvick** [comment]
@DeviaVir that should all be well more than current enough

Let's have you run the make commands manually to see why gitdeploy_git-deploy_1 is not being created for you:

```
docker rm -f gitdeploy_git-deploy_1 || return 0
docker-compose -f docker-compose.yml -f ./test/docker-compose.test.yml down --remove-orphans
docker-compose -f docker-compose.yml -f ./test/docker-compose.test.yml build
docker-compose -f docker-compose.yml -f ./test/docker-compose.test.yml up -d
docker exec -it gitdeploy_test_1 bats test.bats
```

**Comment by DeviaVir** [comment]
First step:
```
 Chase ~ git-deploy $ docker rm -f gitdeploy_git-deploy_1 || return 0
Error response from daemon: No such container: gitdeploy_git-deploy_1
```

File: test/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,29 @@
+FROM debian:jessie
+
+RUN sed -i "s/httpredir.debian/mirrors.kernel/g" /etc/apt/sources.list && \
+    apt-get update && \
+    apt-get install -y --force-yes \
+        wget \
+        ansible \
```
Comment (thepwagner) [inline test/Dockerfile]: Some of this package list is confusing; required by bats install?

File: test/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,29 @@
+FROM debian:jessie
+
+RUN sed -i "s/httpredir.debian/mirrors.kernel/g" /etc/apt/sources.list && \
```
Comment (thepwagner) [inline test/Dockerfile]: Why change this? (fastest for you vs fastest for everyone)

File: test/test.bats [hunk 3]
```diff
@@ -1,42 +1,33 @@
 load test_helper
 
 setup(){
+	rm -rf /tmp/git-deploy-test
 	mkdir -p /tmp/git-deploy-test
-	create_data_volume
+    destroy_backups
```
Comment (thepwagner) [inline test/test.bats]: Tabs vs spaces battle with yourself.

File: test/test.bats [hunk 4]
```diff
@@ -313,31 +276,28 @@ teardown(){
...
-	run_container
 	ssh_command "genkey testkey"
 
 	date | ssh_command "secret testkey"
 	[ "$?" -eq 0 ]
 }
 
 @test "Generate application secret but key not created" {
-	run_container
+	reset_container
```
Comment (thepwagner) [inline test/test.bats]: Weird smell: other tests only `reset_container` after doing something to the container initially.

File: test/test.sh [hunk 5]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
```
Comment (thepwagner) [inline test/test.sh]: This looks identical to `test/test`; do you need both?

File: test/test_helper.bash [hunk 6]
```diff
@@ -1,92 +1,70 @@
...
+		--volumes-from gitdeploy_git-deploy-data_1 \
 		-v /dev/urandom:/dev/random \
 		-p 2222:2222 \
-		pebble/test-git-deploy
-		#pebble/test-git-deploy &> /dev/null
-	sleep 5
-	gen_sshkey
+		--name gitdeploy_git-deploy_1 \
+		gitdeploy_git-deploy
+		#pebble/gitdeploy_git-deploy_1 &> /dev/null
```
Comment (thepwagner) [inline test/test_helper.bash]: Commented out line makes even less sense now.

**Comment by thepwagner** [comment]
The docker commands that manipulate/name things like docker-compose containers feel fragile; can they use docker-compose directly?

File: Makefile [hunk 7]
```diff
@@ -1,13 +1,17 @@
...
-endif
-
 all:
 
-test:
-	docker build -t pebble/test-git-deploy .
-	@ $(ENV_VARS) bats test/test.bats
+clean:
+	docker rm -f gitdeploy_git-deploy_1 || return 0
+	docker-compose -f docker-compose.yml -f ./test/docker-compose.test.yml down --remove-orphans
```
Comment (thepwagner) [inline Makefile]: Can we just merge these `docker-compose.yml` files? It seems like they're only ever referenced in this combination/order

**Comment by lrvick** [comment]
@thepwagner The issue is I need to change environment variables and while there are a bunch of issues on docker to allow redefining env vars on the fly nothing has been implemented yet, thus the hack for now.

File: test/Dockerfile (hunk 2 above)
Comment (lrvick) [inline test/Dockerfile]: httpredir has been unreliable, travis builds failing constantly. However this appears to be a bug in the way the debian base images are built that "apt-get clean" works around.

File: docker-compose.yml [hunk 8]
```diff
@@ -1,20 +1,15 @@
 version: '2'
 services:
   git-deploy:
+    container_name: 'git-deploy'
```
Comment (thepwagner) [inline docker-compose.yml]: Nice!

File: bin/login [hunk 9]
```diff
@@ -1,6 +1,5 @@
 #!/bin/bash
 
-PATH=$PATH:/git/git-shell-commands
 export CURRENT_USER=$1
 
-eval ${SSH_ORIGINAL_COMMAND}
+exec git-shell -c "$SSH_ORIGINAL_COMMAND"
```
Comment (DeviaVir) [inline bin/login]: 🎉

File: test/test.bats [hunk 10]
```diff
@@ -1,40 +1,39 @@
...
+	destroy_backups
+	set_container "git-deploy-test"
 }
 
 teardown(){
-	rm -rf /tmp/git-deploy-test
-	destroy_data_volume
-	destroy_container
+    rm -rf /tmp/git-deploy-test
```
Comment (DeviaVir) [inline test/test.bats]: tabs or spaces?

File: test/test.bats [hunk 11]
```diff
@@ -1,40 +1,39 @@
...
 	run clone_repo testrepo
 	[ "$status" -eq 0 ]
 }
 
 @test "Reject if HOOK_REPO_VERIFY and no known signature on HOOK_REPO" {
-	run_container
-	make_hook_repo testhookrepo
+    set_container "git-deploy-test-exthooks-sig"
+
+    make_hook_repo
```
Comment (DeviaVir) [inline test/test.bats]: tabs or spaces?

File: test/test.bats [hunk 12]
```diff
@@ -63,7 +61,8 @@ teardown(){
 }
 
 @test "Internal pre-receive hook ignored if HOOK_REPO is defined" {
-	run_container /git/testhookrepo
+    set_container "git-deploy-test-exthooks"
```
Comment (DeviaVir) [inline test/test.bats]: tabs or spaces?

File: test/test.bats [hunk 13]
```diff
@@ -72,13 +71,12 @@ teardown(){
 }
 
 @test "External pre-receive hook can reject bad commit" {
-	run_container
-	ssh_command "mkrepo testhookrepo"
+    set_container "git-deploy-test-exthooks"
```
Comment (DeviaVir) [inline test/test.bats]: tabs or spaces?

File: test/test.bats [hunk 14]
```diff
@@ -93,13 +91,11 @@ teardown(){
 }
 
 @test "External pre-receive hook in branch can reject bad commit" {
-	run_container
-	ssh_command "mkrepo testhookrepo"
+    set_container "git-deploy-test-exthooks"
```
Comment (DeviaVir) [inline test/test.bats]: tabs or spaces?

File: test/test.bats [hunk 15]
```diff
@@ -116,13 +112,12 @@ teardown(){
 }
 
 @test "External pre-receive hook can reject bad commit without priming" {
-	run_container
-	ssh_command "mkrepo testhookrepo"
+    set_container "git-deploy-test-exthooks"
```
Comment (DeviaVir) [inline test/test.bats]: tabs or spaces?

File: test/test.bats [hunk 16]
```diff
@@ -142,13 +136,11 @@ teardown(){
 }
 
 @test "External update hook can reject bad commit" {
-	run_container
-	ssh_command "mkrepo testhookrepo"
+    set_container "git-deploy-test-exthooks"
```
Comment (DeviaVir) [inline test/test.bats]: tabs or spaces?

File: test/test.bats [hunk 17]
```diff
@@ -163,13 +155,11 @@ teardown(){
 }
 
 @test "External update hook in branch can reject bad commit" {
-	run_container
-	ssh_command "mkrepo testhookrepo"
+    set_container "git-deploy-test-exthooks"
```
Comment (DeviaVir) [inline test/test.bats]: tabs or spaces?

File: test/test.bats [hunk 18]
```diff
@@ -195,13 +184,11 @@ teardown(){
 }
 
 @test "External post-receive hook can echo text" {
-	run_container
-	ssh_command "mkrepo testhookrepo"
+    set_container "git-deploy-test-exthooks"
```
Comment (DeviaVir) [inline test/test.bats]: tabs or spaces?

File: test/test.bats [hunk 19]
```diff
@@ -210,13 +197,11 @@ teardown(){
 
 
 @test "External post-receive hook in branch can echo text" {
-	run_container
-	ssh_command "mkrepo testhookrepo"
+    set_container "git-deploy-test-exthooks"
```
Comment (DeviaVir) [inline test/test.bats]: tabs or spaces?

File: test/test.bats [hunk 20]
```diff
@@ -225,13 +210,11 @@ teardown(){
 }
 
 @test "Concurrent pre-receive hooks are sandboxed" {
-	run_container
-	ssh_command "mkrepo testhookrepo"
+    set_container "git-deploy-test-exthooks"
```
Comment (DeviaVir) [inline test/test.bats]: tabs or spaces?

File: test/test.bats [hunk 21]
```diff
@@ -250,13 +233,11 @@ teardown(){
 }
 
 @test "Sandbox locks expire" {
-	run_container
-	ssh_command "mkrepo testhookrepo"
+    set_container "git-deploy-test-exthooks"
```
Comment (DeviaVir) [inline test/test.bats]: tabs or spaces?

File: test/test.bats [hunk 22]
```diff
@@ -269,38 +250,29 @@ teardown(){
...
 	[ "$status" -eq 1 ]
 	echo $output | grep -q "Usage"
 }
 
 @test "Generate encryption key callback" {
-	run_container
-
-	ssh_command "mkrepo testhookrepo"
+    set_container "git-deploy-test-exthooks"
+	make_hook_repo
```
Comment (DeviaVir) [inline test/test.bats]: both tabs and spaces?

File: test/test.bats [hunk 23]
```diff
@@ -387,50 +351,42 @@ ${lines[1]}
...
+	run ssh_command "user" test-keys/test-sshkey2
 	[ $status -eq 0 ]
 }
 
 @test "Run script from hook dir" {
-	run_container
-	ssh_command "mkrepo testhookrepo"
-	ssh_command "mkrepo testrepo.git"
-	clone_repo testhookrepo
+    set_container "git-deploy-test-exthooks"
```
Comment (DeviaVir) [inline test/test.bats]: tabs or spaces?

File: test/test.bats [hunk 24]
```diff
@@ -440,13 +396,11 @@ ${lines[1]}
 }
 
 @test "Run script from hook dir - quotes" {
-	run_container
-	ssh_command "mkrepo testhookrepo"
-	ssh_command "mkrepo testrepo.git"
-	clone_repo testhookrepo
+    set_container "git-deploy-test-exthooks"
```
Comment (DeviaVir) [inline test/test.bats]: tabs or spaces?

File: test/test.bats [hunk 25]
```diff
@@ -457,37 +411,34 @@ ${lines[1]}
 }
 
 @test "Run script from hook dir - config.env" {
-	run_container
-	ssh_command "mkrepo testhookrepo"
-	ssh_command "mkrepo testrepo.git"
-	clone_repo testhookrepo
+    set_container "git-deploy-test-exthooks"
```
Comment (DeviaVir) [inline test/test.bats]: tabs or spaces?

File: test/test.bats [hunk 26]
```diff
@@ -457,37 +411,34 @@ ${lines[1]}
...
 	run ssh_command "run testrepo hello World"
 	[ $status -eq 1 ]
 }
 
 @test "Run script from hook dir - no pull" {
-	run_container
-	ssh_command "mkrepo testhookrepo"
-	ssh_command "mkrepo testrepo.git"
-	clone_repo testhookrepo
+    set_container "git-deploy-test-exthooks"
```
Comment (DeviaVir) [inline test/test.bats]: tabs or spaces?

**Comment by lrvick** [comment]
All tests pass locally again, and fully using docker-compose workflow here without the dirty replacement hacks. And yet travis hates it again...

**Comment by thepwagner** [comment]
+1

**Comment by DeviaVir** [comment]
+1

**Comment by thepwagner** [comment]
+1; glorious tabs!

**Comment by thepwagner** [comment]
@lrvick anything blocking merge?

**Comment by lrvick** [comment]
@thepwagner nope. merging.

**Pull request merged**